cas-server-core-web/src/main/java/org/jasig/cas/web/flow/AbstractNonInteractiveCredentialsAction.java

```diff
@@ -68,17 +78,21 @@ protected final Event doExecute(final RequestContext context) {
         final String ticketGrantingTicketId = WebUtils.getTicketGrantingTicketId(context);
         final Service service = WebUtils.getService(context);
 
-        if (isRenewPresent(context)
-            && ticketGrantingTicketId != null
-            && service != null) {
+        if (isRenewPresent(context) && ticketGrantingTicketId != null && service != null) {
 
             try {
+                final AuthenticationContextBuilder builder = new DefaultAuthenticationContextBuilder(
```

Comment:
Here is another pass at the API based on the collective feedback:

```java
final AuthenticationContextBuilder builder = new DefaultAuthenticationContextBuilder(
      this.authenticationSystemSupport.getPrincipalElectionStrategy());
final AuthenticationTransaction transaction = AuthenticationTransaction.wrap(new ClientCredential(credentials));
this.authenticationSystemSupport.getAuthenticationTransactionManager().handle(transaction,  builder);
final AuthenticationContext authenticationContext = builder.build(service);
final TicketGrantingTicket tgt = this.centralAuthenticationService.createTicketGrantingTicket(authenticationContext);
```

- I removed the `AuthenticationTransactionFactory` and added a factory method to wrap credentials. I would have opted for `for`, except that I can't seeing how it's a keyword. 
- The sanitation of credentials is now moved to the `AuthenticationTransaction`, since the factory is gone.
- As discussed, since we need the Service in the context, I opted to keep the `AuthenticationContext` around and the builder that goes along with that as the carrier.
- I also do agree with Dima, that I wouldnt want upper layers to have access to the `AuthenticationManager`. The TransactionManager is the right intermediate glue here.

Note that this API does not have to be "perfect". It has to be functional. It has to support MFA. It can always be revved later as we don't really consider any of this an "extension point".